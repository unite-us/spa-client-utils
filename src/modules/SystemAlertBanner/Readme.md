### SystemAlertBanner
```jsx

<div>
  <SystemAlertBanner
    text="Word Up!" 
    currentUser={{}}
    messageId="1"
  />
</div>
```
