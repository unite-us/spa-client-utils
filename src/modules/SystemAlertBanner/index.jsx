import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Icon } from '@unite-us/ui';

class SystemAlertBanner extends Component {
  constructor(props) {
    super(props);

    this.hideSystemAlertBanner = this.hideSystemAlertBanner.bind(this);

    this.state = {
      hideBanner: false,
    };
  }

  componentDidMount() {
    const {
      currentUser,
      messageId,
    } = this.props;

    const { id } = currentUser;

    // grab systemAlerts out of localStorage
    const systemAlerts = localStorage.getItem('systemAlerts');

    if (systemAlerts) {
      // don't try to parse a null, bad things happen
      const parsedAlerts = JSON.parse(systemAlerts);

      // check to see if there is a key for the current user
      // in systemAlerts that contains the message id
      if (parsedAlerts[id] && parsedAlerts[id].includes(messageId)) {
        // hide the message

        this.hideSystemAlertBanner();
      }
    }
  }

  hideSystemAlertBanner() {
    const {
      currentUser,
      messageId,
    } = this.props;

    const { id } = currentUser;

    const systemAlerts = localStorage.getItem('systemAlerts');
    // get existing systemAlerts out of localStorage and parse if exists

    const parsedAlerts = systemAlerts ? JSON.parse(systemAlerts) : {};

    // get current user's existing hidden system alerts
    const hiddenAlerts = parsedAlerts[id] || [];

    // create new system alerts object including current user and message id
    const newSystemAlerts = {
      ...parsedAlerts,
      [id]: [...new Set([...hiddenAlerts, messageId])],
    };

    // hide the banner
    this.setState({
      hideBanner: true,
    });

    // add message id to systemAlerts and stick back in localStorage
    localStorage.setItem('systemAlerts', JSON.stringify(newSystemAlerts));
  }

  render() {
    const { text } = this.props;
    const { hideBanner } = this.state;

    if (hideBanner) {
      return null;
    }

    return (
      <div className="system-alert-banner">
        <div className="system-alert-banner__icon">
          <Icon
            color="#2C405A"
            icon="IconInfoCircle"
            size={28}
          />
        </div>
        <div className="system-alert-banner__text-container">{text}</div>
        <div className="system-alert-banner__dismiss-container">
          <button
            className="system-alert-banner__dismiss"
            id="system-alert-banner-dismiss-btn"
            type="button"
            onClick={this.hideSystemAlertBanner}
          >
            <Icon
              color="#FFF"
              icon="IconCross"
              size={12}
            />
          </button>
        </div>
      </div>
    );
  }
}

SystemAlertBanner.propTypes = {
  currentUser: PropTypes.object.isRequired,
  messageId: PropTypes.string.isRequired,
  text: PropTypes.node.isRequired,

};

export default SystemAlertBanner;
