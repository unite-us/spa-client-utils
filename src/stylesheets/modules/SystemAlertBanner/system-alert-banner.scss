$system-alert-color: #EC6C23;

.system-alert-banner {
  background-color: $system-alert-color;
  display: flex;
  z-index: z('appbar');
  margin: 0 -15px 0 -15px;
  flex-direction: row;
  flex-wrap: nowrap;

  &__icon {
    flex: 0 0 auto;
    position: relative;
    margin: 10px 15px 0 10px;

    svg {
      background-color: white;
      border-radius: 50%;
    }
  }

  &__text-container {
    flex: 1 1 auto;
    margin: 15px 15px 15px 0;
    color: white;
  }

  &__text,
  &__link {
    color: white;
  }

  &__link:hover {
    color: $display-color;
  }

  &__dismiss-container {
    flex: 0 0 auto;
    position: relative;
  }

  &__dismiss {
    border: none;
    background: none;
    padding: 5px;
    position: relative;
    right: 3px;
    top: 3px;
    bottom: auto;
    width: 40px;
  }
}
