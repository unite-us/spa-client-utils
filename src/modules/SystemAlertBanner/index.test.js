import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import mockStorage from '../../../testUtils/mockStorage';
import SystemAlertBanner from './index';


describe('SystemAlertBanner', () => {
  global.localStorage = mockStorage();

  afterEach(() => {
    global.localStorage.removeItem('systemAlerts');
  });

  const currentUser = {
    id: '1',
  };
  const someMessage = {
    text: 'this is great, really',
    id: 'some-text-1.0.0',
  };


  it('displays a banner', () => {
    const testRenderer = renderer.create(
      <SystemAlertBanner
        text={someMessage.text}
        messageId={someMessage.id}
        currentUser={currentUser}
      />,
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('hides the banner when you click dismiss button', () => {
    const comp = shallow(
      <SystemAlertBanner
        text={someMessage.text}
        messageId={someMessage.id}
        currentUser={currentUser}
      />,
    );

    comp.find('#system-alert-banner-dismiss-btn').simulate('click');


    expect(comp.find('.system-alert-banner')).toHaveLength(0);
  });

  it('stores message id and user id in a systemAlerts value in localStorage', async () => {
    const comp = shallow(
      <SystemAlertBanner
        text={someMessage.text}
        messageId={someMessage.id}
        currentUser={currentUser}
      />,
    );

    const alerts = global.localStorage.getItem('systemAlerts');
    const parsedAlerts = JSON.parse(alerts);

    // make sure we are starting off with a null
    expect(parsedAlerts).toBeNull();

    // dismiss the banner
    await comp.find('#system-alert-banner-dismiss-btn').simulate('click');

    const newAlerts = global.localStorage.getItem('systemAlerts');
    const parsedNewAlerts = JSON.parse(newAlerts);

    // expect systemAlerts to be in localStorage
    expect(parsedNewAlerts).toEqual({
      1: ['some-text-1.0.0'],
    });
  });

  it('stores message id and user id in a systemAlerts value in localStorage', async () => {
    const comp = shallow(
      <SystemAlertBanner
        text={someMessage.text}
        messageId={someMessage.id}
        currentUser={currentUser}
      />,
    );
    const newSystemAlerts = {
      1: ['some-text-1.0.0.0'],
    };

    localStorage.setItem('systemAlerts', JSON.stringify(newSystemAlerts));

    const alerts = global.localStorage.getItem('systemAlerts');
    const parsedAlerts = JSON.parse(alerts);

    expect(parsedAlerts).toEqual({
      1: [
        'some-text-1.0.0.0',
      ],
    });


    localStorage.setItem('systemAlerts', JSON.stringify(newSystemAlerts));
    // dismiss the banner
    await comp.find('#system-alert-banner-dismiss-btn').simulate('click');

    const newAlerts = global.localStorage.getItem('systemAlerts');
    const parsedNewAlerts = JSON.parse(newAlerts);

    // expect systemAlerts to be in localStorage
    expect(parsedNewAlerts).toEqual({
      1: [
        'some-text-1.0.0.0',
        'some-text-1.0.0',
      ],
    });
  });
});
