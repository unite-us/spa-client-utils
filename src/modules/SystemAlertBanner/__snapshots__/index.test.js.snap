// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SystemAlertBanner displays a banner 1`] = `
<div
  className="system-alert-banner"
>
  <div
    className="system-alert-banner__icon"
  >
    <svg
      className="ui-icon"
      height="28px"
      id=""
      style={
        Object {
          "display": "inline-block",
          "fill": "#2C405A",
          "verticalAlign": "middle",
        }
      }
      viewBox="0 0 27 32"
      width="28px"
    >
      <path
        d="M18.286 24.571v-2.857q0-0.25-0.161-0.411t-0.411-0.161h-1.714v-9.143q0-0.25-0.161-0.411t-0.411-0.161h-5.714q-0.25 0-0.411 0.161t-0.161 0.411v2.857q0 0.25 0.161 0.411t0.411 0.161h1.714v5.714h-1.714q-0.25 0-0.411 0.161t-0.161 0.411v2.857q0 0.25 0.161 0.411t0.411 0.161h8q0.25 0 0.411-0.161t0.161-0.411zM16 8.571v-2.857q0-0.25-0.161-0.411t-0.411-0.161h-3.429q-0.25 0-0.411 0.161t-0.161 0.411v2.857q0 0.25 0.161 0.411t0.411 0.161h3.429q0.25 0 0.411-0.161t0.161-0.411zM27.429 16q0 3.732-1.839 6.884t-4.991 4.991-6.884 1.839-6.884-1.839-4.991-4.991-1.839-6.884 1.839-6.884 4.991-4.991 6.884-1.839 6.884 1.839 4.991 4.991 1.839 6.884z"
      />
    </svg>
  </div>
  <div
    className="system-alert-banner__text-container"
  >
    this is great, really
  </div>
  <div
    className="system-alert-banner__dismiss-container"
  >
    <button
      className="system-alert-banner__dismiss"
      id="system-alert-banner-dismiss-btn"
      onClick={[Function]}
      type="button"
    >
      <svg
        className="ui-icon"
        height="12px"
        id=""
        style={
          Object {
            "display": "inline-block",
            "fill": "#FFF",
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 21 21"
        width="12px"
      >
        <g>
          <path
            d="M2.592.044l18.364 18.364-2.548 2.548L.044 2.592z"
          />
          <path
            d="M0 18.364L18.364 0l2.548 2.548L2.548 20.912z"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;
